### PR TITLE
DAOS-12287 test: CR Pass 4 - Orphan container

### PR DESCRIPTION
**Pull request by shimizukko**
Test steps:
1. Create a pool and a container.
2. Inject fault to cause orphan container. i.e., container is left in the system, but doesn't appear with daos commands.
3. Check that the container doesn't appear with daos command.
4. Stop servers.
5. Use ddb to verify that the container is left in shards.
6. Enable the checker.
7. Set policy to --all-interactive.
8. Start the checker and query the checker until the fault is detected.
9. Repair by selecting the destroy option.
10. Query the checker until the fault is repaired.
11. Disable the checker.
12. Run the ddb command and verify that the container is removed from shard.

Create recovery_test_base.py to store methods used across the CR tests.

Add daos faults command support in daos_utils_base.py and daos_utils.py

Skip-unit-tests: true
Skip-fault-injection-test: true
Test-tag: test_orphan_container ddb_cmd
Test-repeat: 4
Required-githooks: true

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

File: src/tests/ftest/util/recovery_test_base.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+        """
+        hosts = NodeSet(self.hostlist_servers[0])
+        scm_mount = self.server_managers[0].get_config_value("scm_mount")
+        vos_path = os.path.join(scm_mount, pool.uuid.lower())
+        command = " ".join(["sudo", "ls", vos_path])
+        cmd_out = run_pcmd(hosts=hosts, command=command)
+
+        # return vos_file
+        for file in cmd_out[0]["stdout"]:
```
Comment (daltonbohning) [inline src/tests/ftest/util/recovery_test_base.py:42]: It's better to use run_remote than run_pcmd. Something like this...
```suggestion
        from run_utils import run_remote
        cmd_out = run_remote(self.log, hosts, command)

        # return vos_file
        for file in cmd_out.output[0].stdout:
```

File: src/tests/ftest/recovery/container_list_consolidation.py [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+        ls_out = "\n".join(cmd_result[0]["stdout"])
+        uuid_regex = r"([0-f]{8}-[0-f]{4}-[0-f]{4}-[0-f]{4}-[0-f]{12})"
+        match = re.search(uuid_regex, ls_out)
+        if match:
+            errors.append("Container UUID is found in shard! Checker didn't remove it.")
+
+        # Start server to prepare for the cleanup.
+        dmg_command.system_start()
+
+        report_errors(test=self, errors=errors)
```
Comment (daltonbohning) [inline src/tests/ftest/recovery/container_list_consolidation.py:173]: Just pointing out that if system_start raises an exception and causes the test to fail, then we would miss the report_errors.

**Comment by shimizukko** [comment]
Blocked due to https://daosio.atlassian.net/browse/DAOS-14473 (Engine crash during `dmg check query` after the repair.)

File: src/tests/ftest/util/recovery_test_base.py (hunk 1 above)
Comment (shimizukko) [reply src/tests/ftest/util/recovery_test_base.py]: Done. Thanks.

File: src/tests/ftest/recovery/container_list_consolidation.py (hunk 2 above)
Comment (shimizukko) [reply src/tests/ftest/recovery/container_list_consolidation.py]: Added `except CommandFailure` to catch the system start exception.

**Review by daltonbohning — APPROVED** [review]

**Review by dinghwah — APPROVED** [review]

**Comment by shimizukko** [comment]
The 8 CR test failures (CR20 to CR27) in daos_test/suite.py is unexpected and Fan Yong is looking into them in DAOS-14513.

**Pull request merged**